valhalla/meili/geometry_helpers.h

```diff
@@ -18,24 +19,23 @@ namespace helpers {
 // snapped point, sqaured distance, segment index, offset
 inline std::tuple<midgard::PointLL, float, typename std::vector<midgard::PointLL>::size_type, float>
 Project(const midgard::projector_t& p,
-        const typename std::vector<midgard::PointLL>& shape,
+        midgard::Shape7Decoder<midgard::PointLL>& shape,
         float snap_distance = 0.f) {
-  if (shape.empty()) {
-    throw std::invalid_argument("got empty shape");
-  }
-
-  midgard::PointLL closest_point(shape.front());
+  midgard::PointLL first_point(shape.pop());
+  auto closest_point = first_point;
+  auto closest_segment_point = first_point;
   float closest_distance = p.approx.DistanceSquared(closest_point);
-  decltype(shape.size()) closest_segment = 0;
+  size_t closest_segment = 0;
   float closest_partial_length = 0.f;
   float total_length = 0.f;
 
   // for each segment
-  for (decltype(shape.size()) i = 0; i < shape.size() - 1; ++i) {
+  auto u = first_point;
+  size_t i = 0;
+  for (; !shape.empty(); ++i) {
     // project a onto b where b is the origin vector representing this segment
     // and a is the origin vector to the point we are projecting, (a.b/b.b)*b
-    const auto& u = shape[i];
-    const auto& v = shape[i + 1];
+    auto v = shape.pop();
```

Comment:
we decode one point at a time and completely avoid any allocations for shape decoding this gives decent speed up in loki and should do here as well (especially where we do many more checks with lots more input)